Add support for LWJGL3 inside an AWT Canvas

tisoft opened the pull request:
Currently using a jitpack version of https://github.com/LWJGLX/lwjgl3-awt since MacOS Support is not released yet.

Work in Progress! Just opened the PR to be able to discuss issues I encounter.

Fixes #6190

crykn commented on extensions/gdx-tools/src/com/badlogic/gdx/tiledmappacker/TiledMapPacker.java:
```diff
@@ -480,12 +480,11 @@ public static void main (String[] args) {
 		}
 
 		TiledMapPacker packer = new TiledMapPacker(packerSettings);
```
The documentation of `TiledMapPacker#main()` (lines 128-134) needs to be adapted as well.

crykn commented on extensions/gdx-tools/src/com/badlogic/gdx/tools/particleeditor/ParticleEditor.java:
```diff
@@ -428,6 +428,11 @@ public void render () {
 			Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
 
+			//TODO: lwjgl3 currently calls the render method too early
+			if(previewImagePanel==null){
```
Formatting!

tisoft replied:
I hope I can get rid of this again, I think there is something wrong in the initialization order currently

crykn replied:
I found a fix for this: https://github.com/tisoft/libgdx/pull/1.

tisoft replied:
Thanks!

I'll try to make the Lwjgl3AWTCanvas a drop in replacement for the Lwjgl2 stuff, so I hope I can get around the issue, without changing the ParticleEditor. But if I don't manage to do that, I'll just take your PR. 😄

crykn replied:
Sounds good 👍

tisoft replied:
I used your reorder fix now. :)

In Lwjgl2, there are two differnt implementations: `LwjglCanvas` and `LwjglAWTCanvas`. I used the `LwjglAWTCanvas` as base for the new `Lwjgl3AWTCanvas`. The `ParticleEditor` used the `LwjglCanvas`, though. Replacing that with the `LwjglAWTCanvas` gives the same NPE as with my new `Lwjgl3AWTCanvas`. So in deed the `Lwjgl3AWTCanvas` is a drop in replacement for `LwjglAWTCanvas`, but probably not for `LwjglCanvas`.

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTCanvas.java:
```diff
@@ -0,0 +1,541 @@
+		this.listener = listener;
+		if (config == null) config = new Lwjgl3ApplicationConfiguration();
+
+		this.config=config;
```
Formatting

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTCanvas.java:
```diff
@@ -0,0 +1,541 @@
+		if (audio != null) Gdx.audio = audio;
+		if (files != null) Gdx.files = files;
+		if (net != null) Gdx.net = net;
+		Gdx.graphics = graphics;
```
Is redundant.

tisoft replied:
yeah, I copy pasted it from both LwjglCanvas from lwjgl2 and Lwjgl3Window 😆

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTCanvas.java:
```diff
@@ -0,0 +1,541 @@
+			canvas.swapBuffers();
+		}
+
+//		Display.sync(getFrameRate() * instanceCount);
```
You should be able to do that with the `Sync` class in the lwjgl3 backend.

tisoft replied:
I'll try that, thanks!

tisoft replied:
Works (after I initialized GLFW)

crykn reviewed (changes requested):
Hey @tisoft, thanks for your PR! I just had a look at your changes and tested them on MacOS:

- `ParticleEditor` does render its stuff, which is good. However,  if another window is focused (i.e. clicked) before ParticleEditor fully opens, the lwjgl panel is rendered at the wrong location (top left).
- Rendering in `Hiero` works as well, but the lwjgl rendering always happens at the wrong position (maybe because of its splash screen?)
- `Flame` renders its output as expected, but the lwjgl panel is in the top left, so the wrong position issue can’t happen here.
- I didn’t test `TiledMapPacker`, `TiledMapPackerTestRender` and `KTXProcessor`, as I don’t have any experience with them
- TexturePackerTest throws the following error:  `Assertion failed: (shape[0] == SHAPE_START), function assert_check_shape, file /AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Sources/CoreGraphics/CoreGraphics-1355.22/CoreGraphics/Regions/shape.c, line 161.`

Apart from that, I had some minor nitpicks & added some comments to your code, see the review for details.

tisoft commented on backends/gdx-backend-lwjgl3/build.gradle:
```diff
@@ -19,6 +19,7 @@ targetCompatibility = 1.8
 
 dependencies {
     compile libraries.lwjgl3
+    compile ('com.github.tisoft:lwjgl3-awt:727caee400')
```
Since we don't need the Vulkan support, we could exclude it

tisoft replied:
I have added a PR to lwjgl3-awt to mark both lwjgl-opengl and lwjgl-vulkan as optional in the pom. This would fix the dependency-bloat issue.

https://github.com/LWJGLX/lwjgl3-awt/pull/27

Maybe you want to support that PR. 😄

tisoft replied:
Haha, already merged 👯

tisoft replied:
And a release was cut, so I will move to the released version, as soon as it appears in the repos

crykn replied:
Nice 👍

tisoft replied:
release cutting seems to have failed, but we have a snapshot version in OSS sonatype now.

tisoft commented on build.gradle:
```diff
@@ -35,6 +35,7 @@ allprojects {
         jcenter()
         mavenCentral()
         maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
+        maven { url 'https://jitpack.io' }
```
I had it there before, but the dependent modules wouldn't build then anymore, but maybe I did something wrong, will try again 😄

tisoft commented on extensions/gdx-tools/build.gradle:
```diff
@@ -14,9 +14,12 @@
  * limitations under the License.
  ******************************************************************************/
 
+sourceCompatibility = 1.8
```
For me it wouldn't build without all the 1.7->1.8 changes, I will have a look though, if they are really needed, or if I can get rid of them.

SimonIT replied:
I think it's needed:
https://github.com/LWJGLX/lwjgl3-awt/blob/05784f9517ac3fa6ce30fa17a2b61c5c2d10ee3e/pom.xml#L39-L40

tisoft replied:
Yes, it is. In fact, the whole Lwjgl 3 is Java 8: https://github.com/LWJGL/lwjgl3/issues/177

tisoft replied:
See https://github.com/libgdx/libgdx/pull/6247#issuecomment-732829622

crykn commented:
I just experimented a bit with ParticleEditor: moving `setVisible(true);`out of the `EventQueue.invokeLater()` block (like [here](https://github.com/crykn/libgdx/blob/49f3e4614808752781e62ef333a6f2bfa8ad494a/extensions/gdx-tools/src/com/badlogic/gdx/tools/particleeditor/ParticleEditor.java#L89-L109)) seems to decrease the chances of the positioning issue happening considerably.

tisoft commented:
> * TexturePackerTest throws the following error:    `Assertion failed: (shape[0] == SHAPE_START), function assert_check_shape, file /AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Sources/CoreGraphics/CoreGraphics-1355.22/CoreGraphics/Regions/shape.c, line 161.`

This is caused by a missing -XstartOnFirstThread, that Lwjgl3Application needs. I have replaced it with Lwjgl3AWTFrame now, that doesn't need this parameter, since AWT is doing the correct stuff there for us.

I don't know what all this tools are supposed to do, so I haven't tried them, though. 🙈

tisoft commented:
> * `ParticleEditor` does render its stuff, which is good. However,  if another window is focused (i.e. clicked) before ParticleEditor fully opens, the lwjgl panel is rendered at the wrong location (top left).
> * Rendering in `Hiero` works as well, but the lwjgl rendering always happens at the wrong position (maybe because of its splash screen?)
> * `Flame` renders its output as expected, but the lwjgl panel is in the top left, so the wrong position issue can’t happen here.

The wrong position should be fixed now.

(Waiting for https://github.com/LWJGLX/lwjgl3-awt/pull/34 to be merged, currently using a jitpack build.)

tisoft commented:
@crykn I have checked the 1.8 language level again.

* gdx-tools needs 1.8, since it is now based on lwjgl3, which needs 1.8
* gdx-tests has a dependency on gdx-tools, for the `BitmapFontWriterTest`
* the other test projects depend on gdx-tests

So if we could move BitmapFontWriterTest somewhere else, most of the language levels could stay on 1.7. I don't know if and where to move it though. :shrug: 

I have added commit 36d51e0003ce2909d2a0b6ae13926dc389c82bb9 to show how it would look like, if the `BitmapFontWriterTest` would be removed.

tisoft commented on tests/gdx-tests-iosrobovm/build.gradle:
```diff
@@ -22,8 +22,8 @@ buildscript {
 
-sourceCompatibility = '1.7'
-targetCompatibility = '1.7'
+sourceCompatibility = '1.8'
```
See https://github.com/libgdx/libgdx/pull/6247#issuecomment-732829622

tisoft commented on tests/build.gradle:
```diff
@@ -17,8 +17,8 @@
 
-    sourceCompatibility = 1.7
-    targetCompatibility = 1.7
+    sourceCompatibility = 1.8
```
See https://github.com/libgdx/libgdx/pull/6247#issuecomment-732829622

crykn commented:
Hey @tisoft! I haven't had time yet to look at your recent changes, but they seem rather promising! Regarding the language levels:

- We can't avoid raising `gdx-tools`' source level when using LWJGL3, so that is fine. Also, if any third-party projects are depending on gdx-tools those should be desktop projects anyway and on desktop Java 8 is no problem.
- Changing the source level of `gdx-tests` looks fine to me as well, because no end-user should depend on those projects. And I'd like to avoid moving BitmapFontWriterTest, since at the moment _all_ tests are in gdx-tests.
- However, I don't know whether the Android & iOS tests are affected by changing the source level (maybe @PokeMMO, @obigu or any of the others could chime in on this). But they seem to compile, so I don't think there'll be any issues.

tisoft commented:
Since everything now works for me ™️ , I thing this is ready for review.

It's not ready for merge until the lwjgl3-awt project has merged https://github.com/LWJGLX/lwjgl3-awt/pull/34

I tested this on Linux and macOS 10.15 with different java versions, but would really appreciate more testing.

obigu commented:
> * However, I don't know whether the Android & iOS tests are affected by changing the source level (maybe @PokeMMO, @obigu or any of the others could chime in on this). But they seem to compile, so I don't think there'll be any issues.

I'm not 100% sure of the consequences of setting source level to 8 on android-tests or ios-tests tests while having their platform dependencies on 6, my instinct says it's not a good idea and that we may regret it in the future. Android and Java 8 have some peculiarities and given that we have min SDK 14 we may come across funny lint errors. If the reason for this is just 1 test, I'd rather temporarily not support that test on Android and iOS (until we move to Java 8).

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTCanvas.java:
```diff
@@ -0,0 +1,564 @@
+	Cursor cursor;
+	private final Sync sync;
+
+	public Lwjgl3ApplicationConfiguration getConfig() {
```
This method shouldn't be before the constructors.

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTCanvas.java:
```diff
@@ -0,0 +1,564 @@
+		if (!running) return;
+		if (!SwingUtilities.isEventDispatchThread()) {
+			//This method must be called on the EDT
+			SwingUtilities.invokeLater(new Runnable() {
```
I think a lambda would be easier to read.

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTCanvas.java:
```diff
@@ -0,0 +1,564 @@
+		Array<LifecycleListener> listeners = lifecycleListeners;
+
+		// To allow destroying of OpenGL textures during disposal.
+		canvas.runInContext(new Runnable() {
```
Same here, but it's all up to you.

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTGraphics.java:
```diff
@@ -0,0 +1,525 @@
+	private long frameCounterStart = 0;
+	private int frames;
+	private int fps;
+	private int windowPosXBeforeFullscreen;
```
These 5 lines can be removed.

crykn commented on backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3AWTGraphics.java:
```diff
@@ -0,0 +1,525 @@
+	private void storeCurrentWindowPositionAndDisplayMode() {
+//		windowPosXBeforeFullscreen = window.getPositionX();
+//		windowPosYBeforeFullscreen = window.getPositionY();
+		windowWidthBeforeFullscreen = logicalWidth;
```
These three lines can be commented out as well.

crykn reviewed (changes requested):
Hey @tisoft, sorry for taking so long to review your recent changes. Your PR looks very good! I have only added some minor requests for changes.  

- ParticleEditor, Hiero and Flame now work flawlessly on Mac OS, as was the original aim of porting the Lwjgl2AWTCanvas to LWJGL3.
- I couldn't really test `TiledMapPacker`, etc., as I don’t have any experience with them. I would appreciate if anyone of the other team members, who have more knowledge of these classes, could take a look.
- There are some minor formatting mistakes, just re-run the formatter once more.
- I noticed one minor "problem": when I close one of the test windows, which were started by `Lwjgl3AwtTestStarter`, then the whole test starter closes.
- A small reminder (not necessarily to you @tisoft): we shouldn't forget to add the new keyboard constants to `Lwjgl3AwtInput`, as soon as @MrStahlfelge's PRs are merged.
- Regarding `BitmapFontWriterTest`: In accordance with @obigu's comment, I would suggest moving it to gdx-tests-lwjgl3, so we can leave the source levels of the other tests unchanged.

tisoft commented:
@crykn I think I fixed all review comments. Only missing is the merge and release of https://github.com/LWJGLX/lwjgl3-awt/pull/34

tisoft commented:
Lwjgl3-awt 0.1.8 has been released, so from my point of view, this could be merged now. 🥳

crykn commented on fetch.xml:
```diff
@@ -8,7 +8,11 @@
     <property name="lwjgl" value="https://oss.sonatype.org/content/repositories/releases/org/lwjgl"/>
     <property name="lwjgl-version" value="3.2.3"/>
-    
+
```
Here is still some Jitpack stuff left.

crykn commented:
Hey @tisoft,

last nitpick, I promise: could you add key codes for `CAPS_LOCK`, `PAUSE` (aka Break), `PRINT_SCREEN`, `SCROLL_LOCK`, `F13` to `F24`, `NUMPAD_DIVIDE`, `NUMPAD_MULTIPLY`, `NUMPAD_SUBTRACT`, `NUMPAD_ADD`, `NUMPAD_DOT`, `NUMPAD_COMMA`, `NUMPAD_ENTER`, `NUMPAD_EQUALS`,  `NUM_LOCK` and `NUMPAD_LEFT_PAREN`, `NUMPAD_RIGHT_PAREN` to Lwjgl3AwtInput? Those were recently introduced with libGDX 1.9.13. Basically, you need to port the changes made to LwjglAwtInput in these three commits (2437e2afe35337e6f7772321e29ca06e1f407231, 325c6386328a3066bed47cc621bc8b993d4e2c7c, 4648b1ca6d63e7b88b6ec7c48ab59ca7f279161f) to Lwjgl3AwtInput.

Apart from that, as we already talked about, I'm not really sure about the `fetch.xml` changes, as build stuff isn't my forte, but maybe one of the others can chime in here.

Otherwise, I think this is good to merge. I wasn't able to test the functionality of all classes in gdx-tools (looking at you `TiledMapPacker`), because I'm not really familiar with their workings in the first place, but all the visual tools started up and rendered correctly. 

**Thanks a lot for all the work you put into this! It is very nice to see macOS support for Hiero, Particle Editor, etc. to be finally back.**

tisoft commented:
@crykn I have added the key bindings, rebased and extended AbstractGraphics.

Should be ready now. Might need a little squashing though. :smiling_imp:

tisoft commented:
- rebased to curent master
- extended AbstractInput
- squashed commits and splited in two semantically meaningful commits

obigu commented:
First of all thank you for the effort on this PR.

I've tested building and running the tools on Windows. On both Particle Editors (2D and 3D), the preview window (I think it's the canvas) blinks lile crazy (I can provide a video if you need it). I haven't been able to test Hiero because there's currenlty an issue that makes it crash at startup but it's unrelated to this RP.

tisoft commented:
Thanks for testing in windows. Unfortunately I don't have a windows system here, so I can't reproduce the error easily. I'll try it in a VM.

crykn commented:
Ah darn. Now it no longer misbehaves on macOS but does so on Windows, it seems. Looking back; I seem to remember that it didn't flicker on Windows before your positon fix. I somehow missed to test the PR on Windows after this change. 

[This](https://github.com/LWJGLX/lwjgl3-awt/blob/main/test/org/lwjgl/opengl/awt/AWTTest.java) lwjgl3-awt test runs fine on Windows btw, so it seems like the issue is in libGDX code.

The Hiero issue with this PR is actually separate to the one @obigu mentioned (which, on another note, I can't reproduce, so it is probably caused by one of the installed fonts). Hiero crashes with this PR, because "No context is current or a function that is not available in the current context was called." See the full stacktrace below:

<details>
  <summary><b><i>Click to show full log</i></b></summary>

```
> Task :extensions:gdx-tools:Hiero.main() FAILED
FATAL ERROR in native method: Thread[AWT-EventQueue-0,6,main]: No context is current or a function that is not available in the current context was called. The JVM will abort execution.
	at org.lwjgl.opengl.GL11C.nglGenTextures(Native Method)
	at org.lwjgl.opengl.GL11C.glGenTextures(GL11C.java:664)
	at org.lwjgl.opengl.GL11.glGenTextures(GL11.java:2492)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3GL20.glGenTexture(Lwjgl3GL20.java:348)
	at com.badlogic.gdx.graphics.Texture.<init>(Texture.java:142)
	at com.badlogic.gdx.graphics.g2d.PixmapPacker$Page$1.<init>(PixmapPacker.java:504)
	at com.badlogic.gdx.graphics.g2d.PixmapPacker$Page.updateTexture(PixmapPacker.java:504)
	at com.badlogic.gdx.graphics.g2d.PixmapPacker.updatePageTextures(PixmapPacker.java:410)
	at com.badlogic.gdx.graphics.g2d.PixmapPacker.updateTextureRegions(PixmapPacker.java:402)
	at com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator.generateData(FreeTypeFontGenerator.java:464)
	at com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator.generateFont(FreeTypeFontGenerator.java:161)
	at com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator.generateFont(FreeTypeFontGenerator.java:152)
	at com.badlogic.gdx.tools.hiero.unicodefont.UnicodeFont.setRenderType(UnicodeFont.java:647)
	at com.badlogic.gdx.tools.hiero.Hiero.updateFont(Hiero.java:315)
	at com.badlogic.gdx.tools.hiero.Hiero$3.valueChanged(Hiero.java:439)
	at javax.swing.JList.fireSelectionValueChanged(JList.java:1802)
	at javax.swing.JList$ListSelectionHandler.valueChanged(JList.java:1816)
	at javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:184)
 	at javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:164)
	at javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:211)
	at javax.swing.DefaultListSelectionModel.changeSelection(DefaultListSelectionModel.java:405)
	at javax.swing.DefaultListSelectionModel.changeSelection(DefaultListSelectionModel.java:415)
	at javax.swing.DefaultListSelectionModel.setSelectionInterval(DefaultListSelectionModel.java:459)
	at javax.swing.JList.setSelectedIndex(JList.java:2216)
	at javax.swing.JList.setSelectedValue(JList.java:2366)
	at com.badlogic.gdx.tools.hiero.Hiero.<init>(Hiero.java:190)
	at com.badlogic.gdx.tools.hiero.Hiero$21.run(Hiero.java:1591)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
 	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```
</details>

crykn commented:
Hey @tisoft did you get the chance to try Lwjgl3AwtCanvas in a Windows VM?

Duna commented:
Put your lazy ass to work! Don't ask others to contribute, is your module! https://stackoverflow.com/questions/64969034/libgdx-particle-editor-macos-big-sur-not-rendered
What a stupid comment on stackoverflow!

tisoft commented:
I have rebased against current master. Unfortunately I could not get it to work in windows inside a VM, so the flickering is probably still there. Could anyone look into the windows flickering issue?

tisoft commented:
Rebased again against master. I still can't test/debug on windows. Would it be better to split the PR? One PR for the addition of the Lwjgl3AWTCanvas, which would not break anything existing. And another PR, that contains the changes of the tools project, which could possibly break, especially on windows, as I can't test there?

crykn commented:
Hm, I'm not sure. The Windows issue also happens with `Lwjgl3AWTCanvas` itself (for example in `Lwjgl3AWTCanvasTest` and `Lwjgl3AwtTestStarter`), so this wouldn't really help with the underlining "problem" of this PR. But if you want to do it for the sake of better readability, sure, feel free to go ahead.

Frosty-J commented:
I've had a quick test of this on Linux. I see the discussion about flickering above. While I haven't tested on Windows, I can get some flicker on resize, though not enough for it to really be an issue. The bigger issue is I managed to get an old copy stuck in one of my buffers, which could render an application unusable until restarted.

Below is a GIF played at 1/6th speed to clearly demonstrate the issue. It's inside a spoiler thingy because it's quite distracting. The black flicker I'm referring to affects the entire window on frame 6 - other blackness on resize is normal.

<details>
<summary>GIF</summary>

![](https://user-images.githubusercontent.com/60154347/129609260-f196f765-7b52-4b22-973e-7840b6f99308.gif)

</details>

Additionally, closing a test crashes the whole thing, but I expect that's due to it creating an additional window and mightn't be an issue someone would have if their game was developed with AWT from the start. Not many games use multiple windows anyway.

```
X Error of failed request:  RenderBadPicture (invalid Picture parameter)
  Major opcode of failed request:  139 (RENDER)
  Minor opcode of failed request:  7 (RenderFreePicture)
  Picture id in failed request: 0x200004d
  Serial number of failed request:  1178
  Current serial number in output stream:  1185
AL lib: (EE) alc_cleanup: 1 device not closed
```

None of the above issues happen when running a non-AWT LWJGL3 application.

Gleethos commented:
Hey, is this issue still being worked on? It looks like most of the work is already done.

undefinedhuman commented:
Hey, is this feature considered to be merged any time soon? Would love to see an AWT integration for LWJGL3

undefinedhuman commented:
What features/bug fixes are missing for this PR, I might be able to help here :)?

SimonIT commented:
@undefinedhuman There's the flickering issue on Windows (see above comments), and in the meantime, some git conflicts popped up
